### PR TITLE
[wip] initial commit for sync bn plugin

### LINKED ISSUE #4349: Support sync batch norm as a plugin
## 🚀 Feature
Support sync batch norm as a plugin so we don't duplicate code across accelerators

### Motivation

It'll look very similar to this: https://github.com/PyTorchLightning/pytorch-lightning/pull/4285

### Pitch

```
import torch
from pytorch_lightning.core.lightning import LightningModule


class SyncBatchNormPlugin(object):
    """
    Plugin to link a custom sync batchnorm implementation to any arbitrary accelerator.
    Example::
        class MySyncBN(SyncBatchNormPlugin):

            def configure_sync_batchnorm(self, model):
                model = MySyncBNWrapper(model)
                return model
        my_sync_bn = MySyncBN()
        trainer = Trainer(sync_batchnorm=True, plugins=[my_sync_bn])
    """

    def configure_sync_batchnorm(self, model: LightningModule) -> LightningModule:
        """
        By default: adds global batchnorm for a model spread across multiple GPUs and nodes.
        Override this to synchronize batchnorm between specific process groups instead
        of the whole world or use a different sync_bn like `apex`'s version.
        Args:
            model: pointer to current :class:`LightningModule`.
        Return:
            LightningModule with batchnorm layers synchronized between process groups
        """
        model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model, process_group=None)

        return model
```

Do we even need the `sync_batchnorm` flag on the `Trainer` with this plugin?

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #4349

TODO:
- fix tests
- add more tests
- resolve constructor args for extensibility: there are going to be more plugin types added. how can we make changes like this less painful/redundant across accelerators?

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — COMMENTED** [review]
I'd like feedback on how we can handle the constructor better. there are going to be more plugin types added. how can we make changes like this less painful/redundant across accelerators?

- i know this is "messy" because of black formatting but it'd make life so much easier to use that as the standard formatting across lightning so we don't have to keep dealing with it
- i know the torch.distrib imports are duplicated based on https://github.com/PyTorchLightning/pytorch-lightning/pull/4337 so i will need to clean that up here too

**Comment by Borda** [comment]
@ananthsub any update here, can we make it in 1.1?

**Comment by tchaton** [comment]
Hey @ananthsub,

Do you want to continue this PR from master directly and extend accelerator to support a BN plugin ?
Or do you want us to do so ?

Best,
T.C
